File: oneSignalManager.ts
import OneSignal from 'react-onesignal';

export interface OneSignalStatus {
  isInitialized: boolean;
  isPushSupported: boolean;
  permission: boolean;
  subscriptionId: string | null;
  optedOut: boolean;
}

class OneSignalManager {
  private static instance: OneSignalManager;
  private isInitialized = false;
  private appId: string;

  private constructor() {
    this.appId = ((import.meta as unknown as { env?: Record<string, string> }).env?.VITE_ONESIGNAL_APP_ID) || '';
  }

  public static getInstance(): OneSignalManager {
    if (!OneSignalManager.instance) {
      OneSignalManager.instance = new OneSignalManager();
    }
    return OneSignalManager.instance;
  }

  /**
   * Initializes the OneSignal Web SDK
   */
  public async initialize(customAppId?: string): Promise<boolean> {
    if (this.isInitialized) {
      return true;
    }

    const targetAppId = customAppId || this.appId;
    if (!targetAppId || targetAppId === 'YOUR_ONESIGNAL_APP_ID') {
      console.warn('[OneSignal] No valid App ID provided. Push notifications will remain in simulated mode.');
      return false;
    }

    try {
      await OneSignal.init({
        appId: targetAppId,
        allowLocalhostAsSecureOrigin: true,
      });

      this.isInitialized = true;
      console.log('[OneSignal] Successfully initialized with App ID:', targetAppId);

      // Register listeners for permission and subscription changes
      OneSignal.User.PushSubscription.addEventListener('change', (event) => {
        console.log('[OneSignal] Push subscription status changed:', event);
      });

      return true;
    } catch (error) {
      console.error('[OneSignal] Initialization error:', error);
      return false;
    }
  }

  /**
   * Requests push notification permissions from the user
   */
  public async requestPermission(): Promise<boolean> {
    try {
      if (!this.isInitialized) {
        const initialized = await this.initialize();
        if (!initialized) {
          // Fallback to standard browser notification if OneSignal is not yet configured with valid App ID
          if (typeof window !== 'undefined' && 'Notification' in window) {
            const perm = await Notification.requestPermission();
            return perm === 'granted';
          }
          return false;
        }
      }

      await OneSignal.Notifications.requestPermission();
      return OneSignal.Notifications.permission;
    } catch (error) {
      console.error('[OneSignal] Request permission error:', error);
      return false;
    }
  }

  /**
   * Gets current subscription details
   */
  public async getStatus(): Promise<OneSignalStatus> {
    const isPushSupported = typeof window !== 'undefined' && ('PushManager' in window || 'Notification' in window);
    if (!this.isInitialized) {
      const standardPermission = typeof window !== 'undefined' && 'Notification' in window && Notification.permission === 'granted';
      return {
        isInitialized: false,
        isPushSupported,
        permission: standardPermission,
        subscriptionId: null,
        optedOut: false,
      };
    }

    try {
      const permission = OneSignal.Notifications.permission;
      const subscriptionId = OneSignal.User.PushSubscription.id || null;
      const optedOut = Boolean(OneSignal.User.PushSubscription.optedIn === false);

      return {
        isInitialized: this.isInitialized,
        isPushSupported,
        permission,
        subscriptionId,
        optedOut,
      };
    } catch {
      return {
        isInitialized: this.isInitialized,
        isPushSupported,
        permission: false,
        subscriptionId: null,
        optedOut: false,
      };
    }
  }

  /**
   * Tag user with active intention details for segmentation or scheduled messaging
   */
  public async trackIntention(intentionText: string, condition: string): Promise<void> {
    if (!this.isInitialized) return;
    try {
      await OneSignal.User.addTags({
        last_intention: intentionText.slice(0, 50),
        condition_group: condition,
        last_active_timestamp: String(Date.now()),
      });
    } catch (error) {
      console.error('[OneSignal] Failed to update user tags:', error);
    }
  }
}

export const oneSignalManager = OneSignalManager.getInstance();

File: OneSignalManager.swift
//
//  OneSignalManager.swift
//  IntentRecovery
//
//  Created for OneSignal iOS SDK v5.x integration.
//

import Foundation
import UIKit
import OneSignalFramework

/// Centralized manager for handling OneSignal push notifications and user state
@objc final public class OneSignalManager: NSObject {
    
    // MARK: - Singleton
    public static let shared = OneSignalManager()
    
    // MARK: - Properties
    private var isInitialized = false
    private var configuredAppId: String?
    
    private override init() {
        super.init()
    }
    
    // MARK: - Initialization
    
    /// Initializes OneSignal with the specified App ID and launch options.
    /// - Parameters:
    ///   - appId: The OneSignal App ID from dashboard (or fallback to Info.plist value)
    ///   - launchOptions: The launch options dictionary from UIApplicationDelegate
    @objc public func initialize(withAppId appId: String? = nil, launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil) {
        guard !isInitialized else {
            print("[OneSignalManager] Already initialized.")
            return
        }
        
        let targetAppId = appId ?? Bundle.main.object(forInfoDictionaryKey: "OneSignalAppId") as? String ?? "YOUR_ONESIGNAL_APP_ID"
        self.configuredAppId = targetAppId
        
        // 1. Enable verbose logging for debugging during development
        #if DEBUG
        OneSignal.Debug.setLogLevel(.LL_VERBOSE)
        #else
        OneSignal.Debug.setLogLevel(.LL_WARN)
        #endif
        
        // 2. Initialize OneSignal
        OneSignal.initialize(targetAppId, withLaunchOptions: launchOptions)
        
        // 3. Setup notification click listener
        OneSignal.Notifications.addClickListener(self)
        
        // 4. Setup foreground notification display policy listener
        OneSignal.Notifications.addForegroundLifecycleListener(self)
        
        // 5. Setup push subscription observer
        OneSignal.User.pushSubscription.addObserver(self)
        
        isInitialized = true
        print("[OneSignalManager] OneSignal initialized successfully with App ID: \(targetAppId)")
    }
    
    // MARK: - Permissions & Subscriptions
    
    /// Requests push notification authorization with system prompt
    /// - Parameter fallbackToSettings: If true and previously denied, may direct user to Settings
    /// - Parameter completion: Callback with permission result
    @objc public func requestNotificationPermission(fallbackToSettings: Bool = false, completion: ((Bool) -> Void)? = nil) {
        DispatchQueue.main.async {
            OneSignal.Notifications.requestPermission({ accepted in
                print("[OneSignalManager] User push notification permission response: \(accepted)")
                completion?(accepted)
            }, fallbackToSettings: fallbackToSettings)
        }
    }
    
    /// Returns the current push subscription ID if available
    @objc public func getPushSubscriptionId() -> String? {
        return OneSignal.User.pushSubscription.id
    }
    
    /// Returns whether the device is currently opted in to push
    @objc public func isOptedIn() -> Bool {
        return OneSignal.User.pushSubscription.optedIn
    }
    
    // MARK: - User & Tag Management
    
    /// Sets tags to segment users based on their active intention and study condition
    /// - Parameters:
    ///   - intention: The intention text
    ///   - condition: The experiment condition group (A, B, or C)
    @objc public func trackIntention(text: String, condition: String) {
        let tags: [String: String] = [
            "last_intention": String(text.prefix(60)),
            "condition_group": condition,
            "updated_at": String(Int(Date().timeIntervalSince1970))
        ]
        OneSignal.User.addTags(tags)
    }
    
    /// Logs in a user with a unique external user ID
    /// - Parameter externalId: User identifier
    @objc public func setExternalUserId(_ externalId: String) {
        OneSignal.login(externalId)
    }
    
    /// Logs out current user and unlinks external user ID
    @objc public func logoutUser() {
        OneSignal.logout()
    }
}

// MARK: - OSNotificationClickListener
extension OneSignalManager: OSNotificationClickListener {
    public func onClick(event: OSNotificationClickEvent) {
        let notification = event.notification
        print("[OneSignalManager] Push notification clicked: \(notification.notificationId ?? "")")
        
        if let additionalData = notification.additionalData {
            print("[OneSignalManager] Additional data payload: \(additionalData)")
            
            // Post internal notification for deep linking into Intent Recovery UI
            NotificationCenter.default.post(
                name: NSNotification.Name("OneSignalNotificationOpened"),
                object: nil,
                userInfo: additionalData
            )
        }
    }
}

// MARK: - OSNotificationLifecycleListener (Foreground Display)
extension OneSignalManager: OSNotificationLifecycleListener {
    public func onWillDisplay(event: OSNotificationWillDisplayEvent) {
        let notification = event.notification
        print("[OneSignalManager] Notification received in foreground: \(notification.title ?? "")")
        
        // Show banner and play sound even when app is active
        event.preventDefault()
        event.notification.display()
    }
}

// MARK: - Push Subscription Observer
extension OneSignalManager: OSNotificationPermissionObserver {
    public func onNotificationPermissionDidChange(_ permission: Bool) {
        print("[OneSignalManager] Notification permission state updated: \(permission)")
    }
}
